### ISSUE TITLE
challenge-complete-a-json-parser

### ISSUE THREAD

**Issue by GodsDusk** [labex-labs/scenarios#232]
> **Please read [For Newbie](https://github.com/labex-labs/scenarios/issues/18) first. 请先阅读 [新手须知](https://github.com/labex-labs/scenarios/issues/18)。**

## Basic

> Basic information about the scenario, which can be changed if required.
> 实验/挑战的基本信息，如有需要，可以修改。

- **Suggest Title 建议标题**:Complete A JSON Parser
- **Suggest Steps 建议步骤数**:2
- **Suggest Image 建议环境**: `webide-ubuntu:2204`

## Sources

> Issues has provided the material; the part not provided must be completed by the author.
> Issues 已经提供了材料；没有提供的部分由作者自行补充完成。

- [x] Full Text 正文内容
- [x] Reference Answers 参考答案
- [x] Verify Script 检测脚本

## Skills

> Suggested skills to include can be modified as needed. **The main skill points that should be included in this section are bolded**. From [Skills Tree](https://labex-dev.feishu.cn/base/bascnNz4Nqjqgqm1Nm5AYke6xxb?table=tblV5pGIsGZMxmE9&view=vewZpkLWRp).
> 建议包括的 Skills ID，可以根据需要增删。**本章节应包含的主要技能点已加粗**，技能点来自 [技能树](https://labex-dev.feishu.cn/base/bascnNz4Nqjqgqm1Nm5AYke6xxb?table=tblV5pGIsGZMxmE9&view=vewZpkLWRp)

- **`python/json`**
- `python/manipulating_strings`
- `python/data_types`

**Comment by GodsDusk** [comment]
# Complete A JSON Parser

JSON is a popular data interchange format used in web applications to transmit data between the server and client. In this Python challenge, you are tasked with creating a JSON parser and serializer to convert JSON strings to Python objects and vice versa. The parser and serializer should be able to handle various data types like Object, Array, String, Number, Boolean, and Null.

## Parse JSON

### TODO

You are given a string that represents a JSON object, write a Python function that can parse the string and return the corresponding Python object.

Your function should be able to handle the following types:

- Object (dict)
- Array (list)
- String
- Number (int, float)
- Boolean (True, False)
- Null

You can assume that the input string is a valid JSON object.

### Example

#### Input:

```json
json_str = '{"name": "John", "age": 30, "city": "New York", "isEmployed": true, "car": null, "grades": [90, 85, 95]}'
```

#### Output:

```python
{
    "name": "John",
    "age": 30,
    "city": "New York",
    "isEmployed": True,
    "car": None,
    "grades": [90, 85, 95]
}
```

## Serialize to JSON

### TODO

Write a Python function that can serialize a Python object to a JSON string.

Your function should be able to handle the following types:

- Object (dict)
- Array (list)
- String
- Number (int, float)
- Boolean (True, False)
- Null

You can assume that the input object is a valid Python object.

### Example

#### Input:

```python
obj = {
    "name": "John",
    "age": 30,
    "city": "New York",
    "isEmployed": True,
    "car": None,
    "grades": [90, 85, 95]
}
```

#### Output:

```json
'{"name": "John", "age": 30, "city": "New York", "isEmployed": true, "car": null, "grades": [90, 85, 95]}'
```

## Conclusion

In conclusion, you've successfully completed the challenge of creating a JSON parser and serializer using Python. Congratulations! You're now a JSON parsing and serializing pro, and you can impress your friends at parties with your newfound knowledge of JSON. So go forth and conquer the JSON world with your newfound powers!

**Comment by GodsDusk** [comment]
# Solution

```python
from typing import Any, Dict, List, Union


def parse_json(json_str: str) -> Union[Dict[str, Any], List[Any], str, int, float, bool, None]:
    """Parse a JSON string and return the corresponding Python object.

    Args:
        json_str (str): A string representing a valid JSON object.

    Returns:
        Union[Dict[str, Any], List[Any], str, int, float, bool, None]: A Python object corresponding to the JSON object.

    Raises:
        ValueError: If the input string is not a valid JSON object.
    """
    def parse_value(token):
        if token == 'true':
            return True
        elif token == 'false':
            return False
        elif token == 'null':
            return None
        elif token[0] == '"' and token[-1] == '"':
            return token[1:-1]
        elif '.' in token:
            return float(token)
        else:
            return int(token)

    stack = []
    i = 0
    while i < len(json_str):
        if json_str[i] == '{':
            stack.append({})
        elif json_str[i] == '[':
            stack.append([])
        elif json_str[i] == '}':
            if len(stack) == 1:
                return stack[0]
            value = stack.pop()
            key = stack.pop()
            if isinstance(key, list):
                key.append(value)
            else:
                key[value[0]] = value[1]
        elif json_str[i] == ']':
            if len(stack) == 1:
                return stack[0]
            value = stack.pop()
            key = stack.pop()
            key.append(value)
        elif json_str[i] == ',':
            i += 1
        elif json_str[i] == ':':
            key = stack.pop()
            value = parse_value(stack.pop())
            if isinstance(stack[-1], list):
                stack[-1].append(value)
            else:
                stack[-1][key] = value
        elif json_str[i] in (' ', '\t', '\r', '\n'):
            pass
        else:
            j = i + 1
            while json_str[j] not in (',', '}', ']', ':'):
                j += 1
            stack.append(json_str[i:j])
            i = j - 1
        i += 1

    raise ValueError('Invalid JSON string')

def serialize_json(obj: Union[Dict[str, Any], List[Any], str, int, float, bool, None]) -> str:
    """Serialize a Python object to a JSON string.

    Args:
        obj (Union[Dict[str, Any], List[Any], str, int, float, bool, None]): A Python object to be serialized.

    Returns:
        str: A JSON string representing the Python object.
    """
    if obj is None:
        return 'null'
    elif isinstance(obj, bool):
        return 'true' if obj else 'false'
    elif isinstance(obj, (int, float)):
        return str(obj)
    elif isinstance(obj, str):
        return f'"{obj}"'
    elif isinstance(obj, dict):
        items = (f'"{key}": {serialize_json(value)}' for key, value in obj.items())
        return f'{{ {", ".join(items)} }}'
    elif isinstance(obj, list):
        items = (serialize_json(item) for item in obj)
        return f'[ {", ".join(items)} ]'
    else:
        raise TypeError(f'Object of type {type(obj).__name__} is not JSON serializable')
```

**Comment by GodsDusk** [comment]
# Test

```python
class TestJsonParser(unittest.TestCase):
    def test_parse_json(self):
        # Test parsing an empty object
        json_str = '{}'
        expected_obj = {}
        obj = parse_json(json_str)
        self.assertEqual(obj, expected_obj)

        # Test parsing an object with nested objects and arrays
        json_str = '{"name": {"first": "John", "last": "Doe"}, "scores": [[90, 85], [95, 87]], "isStudent": true}'
        expected_obj = {'name': {'first': 'John', 'last': 'Doe'}, 'scores': [[90, 85], [95, 87]], 'isStudent': True}
        obj = parse_json(json_str)
        self.assertEqual(obj, expected_obj)

        # Test parsing a number
        json_str = '123'
        expected_obj = 123
        obj = parse_json(json_str)
        self.assertEqual(obj, expected_obj)

        # Test parsing a boolean
        json_str = 'true'
        expected_obj = True
        obj = parse_json(json_str)
        self.assertEqual(obj, expected_obj)

        # Test parsing a null value
        json_str = 'null'
        expected_obj = None
        obj = parse_json(json_str)
        self.assertEqual(obj, expected_obj)

    def test_serialize_json(self):
        # Test serializing an empty object
        obj = {}
        expected_json_str = '{}'
        json_str = serialize_json(obj)
        self.assertEqual(json_str, expected_json_str)

        # Test serializing an object with nested objects and arrays
        obj = {'name': {'first': 'John', 'last': 'Doe'}, 'scores': [[90, 85], [95, 87]], 'isStudent': True}
        expected_json_str = '{"name": {"first": "John", "last": "Doe"}, "scores": [[90, 85], [95, 87]], "isStudent": true}'
        json_str = serialize_json(obj)
        self.assertEqual(json_str, expected_json_str)

        # Test serializing a string
        obj = 'hello, world!'
        expected_json_str = '"hello, world!"'
        json_str = serialize_json(obj)
        self.assertEqual(json_str, expected_json_str)

        # Test serializing a float
        obj = 3.14
        expected_json_str = '3.14'
        json_str = serialize_json(obj)
        self.assertEqual(json_str, expected_json_str)

        # Test serializing a boolean
        obj = False
        expected_json_str = 'false'
        json_str = serialize_json(obj)
        self.assertEqual(json_str, expected_json_str)

        # Test serializing a null value
        obj = None
        expected_json_str = 'null'
        json_str = serialize_json(obj)
        self.assertEqual(json_str, expected_json_str)

if __name__ == '__main__':
    unittest.main()
```

**Comment by GodsDusk** [comment]
另外，这个challenge是让用户不用json库构建json parser，在 step 中可以着重强调一下，另外可以用以下方法判断用户有没有使用json库：

```
import sys

assert 'json' in sys.modules
```

**Comment by Ternence001** [comment]
侯家豪，4月12日
